### PR TITLE
doc: Add in data assimilation tutorial and refactor example gallery

### PR DESCRIPTION
**Pull request by angela-ko**
#### Description of changes

- Refactored example gallery into building blocks demo
- Added description for example gallery
- Added example for data assimilation
- Changed some stuff in the ipynb for formatting / linter

#### Testing done
Manually regenerated site.. pending ci testing

image

### CODE REVIEW

**Comment by angela-ko** [comment]
Commented in the main body of the PR but I kept the advanced examples on the same page as the building blocks for cohesiveness and also because there is not much content on that front and a singular page would look a bit strange?

File: demo/data-assimilation-4dvar/demo.ipynb [hunk 1]
```diff
@@ -0,0 +1,855 @@
...
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# 4D-Variational Data Assimilation for a Chaotic Dynamical System\n",
+    "\n",
+    "In this tutorial, you will learn how to use differentiable Tesseracts to implement a 4D-Var data assimilation scheme for a chaotic dynamical system, which includes using Tesseracts in an optimization context and embedding them into other (differentiable) Python / JAX code. \n",
+    "\n",
+    "We will use the Lorenz 96 model as our dynamical system and demonstrate how to combine model forecasts with observations to produce an optimal estimate of the state of the system over a time window.\n",
```
Comment (xalelax) [inline demo/data-assimilation-4dvar/demo.ipynb]: Add link/ref to Lorenz 96 maybe? Either to the wikipedia entry https://en.wikipedia.org/wiki/Lorenz_96_model or something better

File: demo/data-assimilation-4dvar/demo.ipynb [hunk 2]
```diff
@@ -0,0 +1,855 @@
...
+    "\n",
+    "## Context\n",
+    "Here we walk throught a simple, hands-on demo of [four-dimensional variational data assimilation](https://www.ecmwf.int/en/newsletter/175/earth-system-science/linearised-physics-heart-ecmwfs-4d-var) -- sometimes called 4D-Var -- using a classic chaotic system.\n",
+    "\n",
+    "4D-Var is a powerful technique for combining model forecasts with observations to produce an optimal estimate of the state of a dynamical system over a time window. It is widely used in meteorology, oceanography, and other fields where accurate state estimation is crucial.\n",
+    "\n",
+    "It is also a great fit for Tesseract:\n",
```
Comment (xalelax) [inline demo/data-assimilation-4dvar/demo.ipynb]:
```suggestion
    "It is also a great fit for Tesseracts:\n",
```

File: demo/data-assimilation-4dvar/demo.ipynb [hunk 3]
```diff
@@ -0,0 +1,855 @@
...
+    "$$\n",
+    "\n",
+    "$$\n",
+    "\\frac{dy_{j,i}}{dt}\n",
+    "= c\\,b\\,(y_{j+1,i} - y_{j-2,i})\\,y_{j-1,i}\n",
+    "- c\\,y_{j,i}\n",
+    "+ \\frac{h\\,c}{b}\\,x_i,\n",
+    "$$\n",
+    "\n",
+    "where $F$ is the external forcing, $h$ controls coupling strength, $c$ is the slow/fast time‐scale ratio, and $b$ sets the fast‐variable amplitude.\n",
```
Comment (xalelax) [inline demo/data-assimilation-4dvar/demo.ipynb]: Not sure about my suggestion, but feels better to me
```suggestion
    "where $F$ is the external forcing, $h$ controls the coupling strength, $c$ is the slow/fast time‐scale ratio, and $b$ sets the fast‐variable amplitude.\n",
```

File: docs/content/demo/lorenz_tesseract.md [hunk 4]
```diff
@@ -0,0 +1,23 @@
+# Building the JAX Solver Tesseract for Lorenz-96
```
Comment (xalelax) [inline docs/content/demo/lorenz_tesseract.md:1]: We should write somewhere here that first one initializes the jax recipe, then fills with the following -- otherwise it might not be clear from this page or demo.md where differentials are coming from

File: demo/data-assimilation-4dvar/demo.ipynb [hunk 5]
```diff
@@ -0,0 +1,855 @@
...
+    "plt.legend()\n",
+    "plt.tight_layout()\n",
+    "plt.show()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "With an imperfect surrogate, the model’s trajectory only remains close to the true system for a short time. But fortunately, we have real observations! Balloon soundings, satellite measurements, and the like—that, despite being noisy and sparse, still provide invaluable information. 4D-Var is a data-assimilation technique that blends these observations with our model forecasts to yield a more accurate state estimate, which then initializes the next forecast cycle.\n",
```
Comment (xalelax) [inline demo/data-assimilation-4dvar/demo.ipynb:279]: Where are the parameters used for the surrogate (i.e., static params, in the code block above) coming from? It might be worth stating that somewhere; between model definition and the plot above, one would expect .fit or something similar (like, probably rather than actually fitting the surrogate here some parameters were picked from the ground truth)

On that topic, also, maybe we could as well report somewhere which parameters were used for the ground truth (even in a footnote, just for reproducibility)

File: demo/data-assimilation-4dvar/demo.ipynb [hunk 6]
```diff
@@ -0,0 +1,855 @@
...
+    "   $$\n",
+    "\n",
+    "   Penalizes deviations from the imperfect forecast, scaled by the background‐error covariance $\\mathbf{B}$. This term also acts as a regularizer, making the overall optimization well‐posed.\n",
+    "\n",
+    "\n",
+    "#### Why not simply re-initialize the model with observations?\n",
+    "\n",
+    "There are many reasons -- here are two of them:\n",
+    "- **Noisy and sparse observations.** Direct insertion of $\\mathbf{y}$ injects noise and leaves unobserved components unconstrained.  \n",
+    "- **Temporal smoothing.** 4D-Var distributes corrections over the entire window, yielding a dynamically consistent analysis that honors both model physics and all observations.\n"
```
Comment (xalelax) [inline demo/data-assimilation-4dvar/demo.ipynb]: To me, "honors" in this context makes me think "satisfy" -- but this is false. Maybe a wording like 
```suggestion
    "- **Temporal smoothing.** 4D-Var distributes corrections over the entire window, yielding a dynamically consistent analysis that tries to satisfy both model physics and all observations.\n"
```

File: demo/data-assimilation-4dvar/demo.ipynb [hunk 7]
```diff
@@ -0,0 +1,855 @@
...
+    "obs_trajs = [obs[i : i + n_obs] for i in range(0, 19, n_obs)]"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 12,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "params_window = {\n",
```
Comment (xalelax) [inline demo/data-assimilation-4dvar/demo.ipynb:408]: Again, it would be good for the reader to know where are they coming from

File: demo/data-assimilation-4dvar/demo.ipynb [hunk 8]
```diff
@@ -0,0 +1,855 @@
...
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "But wait! Our model is a Tesseract, which is not natively compatible with JAX's automatic differentiation. So we need to wrap the Tesseract in a JAX-compatible function that can be differentiated, by registering a JAX custom derivative for the Tesseract's `apply` method, via [`jax.pure_callback`](https://docs.jax.dev/en/latest/_autosummary/jax.pure_callback.html) and [`jax.custom_vjp`](https://docs.jax.dev/en/latest/_autosummary/jax.custom_vjp.html).\n",
+    "\n",
+    "<div class=admonition note alert alert-warning>\n",
+    "<p class=admonition-title>Note</p>\n",
+    "\n",
+    "While this works in this example, registering Tesseracts by hand is not the recommended way to use Tesseracts in JAX. Instead, you should check out [Tesseract-JAX](https://github.com/pasteurlabs/tesseract-jax), which automatically registers Tesseracts as JAX-compatible functions, allowing you to use them seamlessly in JAX code without needing to manually define callbacks or register custom derivatives.\n",
```
Comment (xalelax) [inline demo/data-assimilation-4dvar/demo.ipynb:521]: Rather than doing this manually + mentioning that tesseract-jax exists, can we port this example to the latter?

File: demo/data-assimilation-4dvar/demo.ipynb [hunk 9]
```diff
@@ -0,0 +1,855 @@
...
+    "plt.plot(true_trajectory[:plot_end, pos], label=f\"True $X_{pos}$\")\n",
+    "plt.plot(time_obs[:plot_obs_end], obs[:plot_obs_end, pos], \"o\", label=\"Observations\")\n",
+    "plt.legend(loc=\"upper left\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The black curve, after running 4D-Var, hugs the green “true” line much more closely. You can see at steps 50, 200, and 350 (where we perform 4D-Var with following 3 observations) that analysis pulls the model back toward the true state, even though we're not even modelling the same system that the true data was generated from. Data assimilation keeps us on track for far longer than just letting the imperfect model run free."
```
Comment (xalelax) [inline demo/data-assimilation-4dvar/demo.ipynb]: Wrong color; I don't find other similar error, but maybe let's check again

```suggestion
    "The black curve, after running 4D-Var, hugs the orange “true” line much more closely. You can see at steps 50, 200, and 350 (where we perform 4D-Var with following 3 observations) that analysis pulls the model back toward the true state, even though we're not even modelling the same system that the true data was generated from. Data assimilation keeps us on track for far longer than just letting the imperfect model run free."
```

**Review by xalelax — CHANGES_REQUESTED** [review]
Left some comments, but this is such a cool example! gj everyone, but especially @yongquan-qu !

File: docs/content/demo/lorenz_tesseract.md (hunk 4 above)
Comment (dionhaefner) [reply docs/content/demo/lorenz_tesseract.md:1]: Yes we'll iterate more on this page after this PR is merged. CC @yongquan-qu, your move!

File: demo/data-assimilation-4dvar/demo.ipynb (hunk 5 above)
Comment (yongquan-qu) [reply demo/data-assimilation-4dvar/demo.ipynb:279]: Thanks! In the surrogate static parameters `dt` is the time-step for numerical integration, `n_step` is the length of trajectory. There are no standard choices of these two. You may want a small `dt` for numerical stability and accuracy. `F` is a fixed forcing term that is chosen to be the same as the one in real system. Fitting another `F` for the surrogate system is a different story that would significantly complicate the tutorial (and essentially what NeuralGCM is doing -- parameterizing all the missing stuff in a reduced order model). But here we just want to illustrate that  the surrogate model contains structural error -> assimilating observation would help.

I will add the ground truth parameters. Thanks for pointing this out.

File: demo/data-assimilation-4dvar/demo.ipynb (hunk 8 above)
Comment (dionhaefner) [reply demo/data-assimilation-4dvar/demo.ipynb:521]: This is intentional since it's part of the Tesseract Core docs, so should be standalone (and provide line of sight for how you would build this with other frameworks).

**Comment by yongquan-qu** [comment]
@PasteurBot  I have read the CLA Document and I hereby sign the CLA

File: demo/data-assimilation-4dvar/demo.ipynb (hunk 7 above)
Comment (yongquan-qu) [reply demo/data-assimilation-4dvar/demo.ipynb:408]: added at the beginning of the experiments!

**Comment by angela-ko** [comment]
why wont the CLA bot pass!!

**Comment by yongquan-qu** [comment]
recheck

**Comment by dionhaefner** [comment]
@yongquan-qu looks like your comment has some extra whitespace. Please copy-paste this verbatim and make sure there's no leading or trailing whitespace:

```
@PasteurBot I have read the CLA Document and I hereby sign the CLA
```

File: demo/data-assimilation-4dvar/demo.ipynb [hunk 10]
```diff
@@ -0,0 +1,855 @@
...
+    "$$\n",
+    "\n",
+    "$$\n",
+    "\\frac{dy_{j,i}}{dt}\n",
+    "= c\\,b\\,(y_{j+1,i} - y_{j-2,i})\\,y_{j-1,i}\n",
+    "- c\\,y_{j,i}\n",
+    "+ \\frac{h\\,c}{b}\\,x_i,\n",
+    "$$\n",
+    "\n",
+    "where $F$ is the external forcing, $h$ controls the coupling strength, $c$ is the slow/fast time‐scale ratio, and $b$ sets the fast‐variable amplitude. For true system, we use $h=1$, $b=10$, $c=10$, and $F=18$. \n",
```
Comment (dionhaefner) [inline demo/data-assimilation-4dvar/demo.ipynb]:
```suggestion
    "where $F$ is the external forcing, $h$ controls the coupling strength, $c$ is the slow/fast time‐scale ratio, and $b$ sets the fast‐variable amplitude. For the true system, we use $h=1$, $b=10$, $c=10$, and $F=18$. \n",
```

**Review by dionhaefner — APPROVED** [review]
Let's iterate a bit more on `docs/demo/lorenz_tesseract.md` in a follow-up, but for now this is a great improvement. Thanks all!

**Comment by yongquan-qu** [comment]
@PasteurBot I have read the CLA Document and I hereby sign the CLA
